fastjsonschema/generator.py

```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
                         '{}_value'.format(self._variable),
                         '{}.{{{}_key}}'.format(self._variable_name, self._variable),
                     )
+
+    def generate_additional_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
+        self.l('{variable}_keys = set({variable}.keys())')
+
+        if 'patternProperties' in self._definition:
+            self.l('pattern_keys = set()')
+            with self.l('for key in {variable}_keys:'):
+                for pattern in self._definition['patternProperties'].keys():
+                    with self.l('if globals()["{}_re"].search(key):', pattern):
+                        self.l('pattern_keys.add(key)')
+            self.l('{variable}_keys -= pattern_keys')
+
+        add_prop_definition = self._definition["additionalProperties"]
+        if add_prop_definition:
+            with self.l('for {variable}_key in {variable}_keys:'):
+                with self.l('if {variable}_key not in "{}":', self._definition.get('properties', [])):
```

Comment:
This generates wrong code:

        if data_key not in "{'foo': {'type': 'array', 'maxItems': 3}, 'bar': {'type': 'array'}}":

I think this condition shouldn't even be here as in `{variable}_keys` is only the additional properties already.

Reply:
You are right, I meant to just check if `{variable}_key` was not in the `properties` keys.  I'll fix that.

I do think we need this check though for this test case:

```
$ ./json_schema_test_suite.py JSON-Schema-Test-Suite/tests/draft4/additionalProperties.json 
✘ additionalProperties.json
  ✘ additionalProperties allows a schema which should validate
    ✘ FALSE_NEGATIVE JsonSchemaException no additional properties is valid: data.foo must be boolean
    ✘ FALSE_NEGATIVE JsonSchemaException an additional valid property is valid: data.bar must be boolean

Summary of 2 tests:

Failures:

✘ FALSE_POSITIVE    0   0.0%
✘ FALSE_NEGATIVE    2 100.0%
⚠ UNDEFINED         0   0.0%
                    2 100.0%

Passes:

✔ TRUE_POSITIVE     0   0.0%
✔ TRUE_NEGATIVE     0   0.0%
                    0   0.0%

⛔ Ignored:          0
Coverage:       2/2 100.0%
```